### PR TITLE
CAT-2594 reimplement SetVariableTest

### PR DESCRIPTION
**Pull request by AntiDog**

### CODE REVIEW

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
please run espresso tests

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java [hunk 1]
```diff
@@ -85,17 +76,4 @@ public void whenStartedBrick() {
 		onBrickAtPosition(1).checkShowsText(R.string.brick_wait);
 	}
 
-	public void addBrick(Class<?> brickHeaderClass, int brickCategoryId) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java]: id prefer to leave this here.
- it is only used in this test class, theres already ~100 brick test classes that work without something like this, and once it ends up in a utility method, people will misuse this for adding bricks instead of writing proper setup methods.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java [hunk 2]
```diff
@@ -84,4 +98,36 @@ public static void createUserListFromDataFragment(String userListName, boolean f
...
+
+	public static void deleteItemWithTextFromFormulaEditorDataList(String text) {
+		onView(withId(R.id.formula_editor_keyboard_data))
+				.perform(click());
+		onView(allOf(withText(text), withId(R.id
+				.fragment_formula_editor_datalist_item_name_text_view)))
+				.perform(longClick());
+		onView(withText(R.string.delete))
+				.perform(click());
+		try {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java]: i dont like try / catches in test code. especially in utility methods, where it can potentially hide errors.
another PR.
ive seen something similar also in another PR, and they solved it via additional parameter.

Comment (AntiDog) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java]: ok, then I'll do it via an additional parameter

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java (hunk 1 above)
Comment (AntiDog) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java]: so you are saying its fine to have a duplicate method in my test case and the WhenStartedBrickTest
 
I'm asking this since on my last pull request, which was also about this test case, you told me that I should remove the duplicates from these two so I am confused a little bit. Or do you want me to access the method within the WhenStartedBrickTest from SetVariableTest?

**Comment by 84n4n4** [comment]
test this please

**Comment by AntiDog** [comment]
I just saw that I accidentally pushed DeleteItemOfUserListBrickTest, which causes the test case to fail I will fix this today.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java (hunk 1 above)
Comment (84n4n4) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java]: ah, so you use this one aswell. then its okish... 
i thought theres just one class using it....

please rename to addBrickViaUi or something...

**Comment by cheidenreich** [comment]
test this please

**Comment by cheidenreich** [comment]
please resolve the conflicts and the 9 checkstyle errors

**Comment by AntiDog** [comment]
I think I resolved it. Can someone please test it again?
Is there a way to run checkstyle locally? If so how?

**Comment by robertpainsi** [comment]
> Is there a way to run checkstyle locally? If so how?

https://github.com/Catrobat/Catroid/wiki/Frequently-Asked-Questions-(Developers)#how-to-run-static-code-analyzing-tools-checkstyle-pmd-lint-locally

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java [hunk 3]
```diff
@@ -86,4 +90,19 @@ public static void createUserListFromDataFragment(String userListName, boolean f
 		onView(withText(userListName))
 				.check(matches(isDisplayed()));
 	}
+
+	public static void deleteItemWithTextFromFormulaEditorDataList(String text, Boolean variableIsInUse) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java]: please use a typedev or enum instead of the boolean for the switch.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java [hunk 4]
```diff
@@ -86,4 +90,19 @@ public static void createUserListFromDataFragment(String userListName, boolean f
...
+		onView(withId(R.id.formula_editor_keyboard_data))
+				.perform(click());
+		onView(allOf(withText(text), withId(R.id
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java]: please find a way on how to do this via onData instead of onView to make this more robust.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/SetVariableTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		onBrickAtPosition(0).checkShowsText(R.string.brick_when_started);
+		onBrickAtPosition(1).checkShowsText(R.string.brick_set_variable);
+		onBrickAtPosition(1).onVariableSpinner(R.id.set_variable_spinner)
+				.performNewVariableInitial(userVariableName);
+		onBrickAtPosition(1).onVariableSpinner(R.id.set_variable_spinner)
+				.performNewVariable(secondUserVariableName);
+		onBrickAtPosition(1).onFormulaTextField(R.id.brick_set_variable_edit_text)
+				.perform(click());
+
+		deleteItemWithTextFromFormulaEditorDataList(userVariableName, BrickTestUtils.UserVariableUsage.IN_USE);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/SetVariableTest.java]: theres now a formulaEditorWrapper that does all this stuff...

Comment (AntiDog) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/SetVariableTest.java]: ah OK, thanks a lot for adding these 😃
Now I can finally get rid of deleteItemWithTextFromFormulaEditorDataList which delayed the pull request already several times because it was put into wrong files or had a bad signature etc... (since October)

**Comment by 84n4n4** [comment]
test this please

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java [hunk 6]
```diff
@@ -79,19 +79,19 @@ public void whenStartedBrick() {
...
 
-		addBrick(WaitBrick.class, R.string.category_control);
+		addBrickViaUi(WaitBrick.class, R.string.category_control);
 
 		onBrickAtPosition(0).checkShowsText(R.string.brick_when_started);
 		onBrickAtPosition(1).checkShowsText(R.string.brick_wait);
 	}
 
-	public void addBrick(Class<?> brickHeaderClass, int brickCategoryId) {
+	public static void addBrickViaUi(Class<?> brickHeaderClass, int brickCategoryId) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/WhenStartedBrickTest.java]: please remove the static

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/SetVariableTest.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+import static org.catrobat.catroid.uiespresso.formulaeditor.utils.FormulaEditorDataListWrapper.onDataList;
+import static org.catrobat.catroid.uiespresso.formulaeditor.utils.FormulaEditorWrapper.onFormulaEditor;
+import static org.hamcrest.Matchers.allOf;
+
+@RunWith(AndroidJUnit4.class)
+public class SetVariableTest {
+
+	@Rule
+	public BaseActivityInstrumentationRule<SpriteActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(SpriteActivity.class, true, false);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/SetVariableTest.java]: please state the fragment in the rules constructor, and then use launchActivity() (without the null)

**Comment by 84n4n4** [comment]
Test this please

**Comment by 84n4n4** [comment]
Please run espresso tests

**Comment by 84n4n4** [comment]
Please run espresso tests
